### ISSUE TITLE
retrieving/parsing db and conditionals for db.where

### ISSUE THREAD

**Issue by casefc3s** [twitch-irc/twitch-irc#110]
I would like to keep a local db for my bot to pull from and add to for things such as blacklisted spam, brief chat history (so it can be included in ban logs), etc. Here's an example of what I'm attempting to do with blacklisting:

```javascript
function isBlacklistSpam(str, callback) {
	var spam = false;

	// get the list of blacklisted words, and compare with str
	db.list('blacklist').then(function(result) {
		console.log(result);
		for (i = 0; i < result.length; i++) {
			if (str.indexOf(result[i]['word']) >= 0) {
				spam = true;
			}
		}

		return callback(spam);
	});	
}
```

The blacklisted words are added via mod command:
```javascript
var msg = message.toLowerCase().split(" ");

// blacklisting command for banlist
if (msg[0] === '!blacklist' && isMod(channel, user) && msg.length > 1) {
	db.insert('blacklist', {word: msg[1]}).then(function() {
		console.log("Added " + msg[1] + " to blacklisted words/links.");
		client.say(channel, "Added " + msg[1] + " to blacklisted words/links.");
	});
}
```

And here's the check against if there's blacklisted content in chat:
```javascript
// spam checks
isBlacklistSpam(message.toLowerCase(), function(spam) {
	if (spam) {
		isMod(channel, botname, function(mod) {
			if (mod) {
				client.ban(channel, user.username);
				client.say(channel, "All spammers must die, " + user.username + ".");
			} else {
				console.log(channel + ' | ' + user.username + ' has posted a malicious link.');
			}
		});
	}
});
```

I'm unable to grab the 'word' element of the blacklist database via the means in isBlacklistSpam(), do I need to parse database output in a special way? Or at least, even if I enter a blacklisted word into chat, the bot doesn't resolve it. I'm uncertain of what I'm doing wrong here.

Here's another example of where I'm having issues with database usage, where I'm tracking users and their chat history:

```javascript
// adding users to a db
db.where('userdata', {name: user.username} && {chan: channel}).then(function(result) {
	var results = result['items'];
	console.log(result);
	if (results.length <= 0) {
		//  adding to db with user obj, chat history, etc
		db.insert('userdata', {name: user.username, chan: channel, user_data: user, messages: [message]}).then(function() {
			console.log(colors.yellow("Added userdata for " + user.username + "."));
		});
	} else {
		results = results[0];
		// user exists already in db
		var cid = results['cid'];
		console.log("CID: " + cid);
		var msgs = results['messages'];
		msgs.push(message);
		console.log(msgs);
		// pop excess messages from front for storage
		if (msgs.length > 10) {
			var rem = Math.abs(10-msgs.length);
			msgs.splice(0,rem);
		}
		db.update('userdata', cid, {user_data: user, messages: msgs}).then(function() {
			console.log(console.yellow("Updated data for " + user.username + "."));
		});
	}
});
```

There's a few issues above, although I've changed things up to try and get it to work so many times I'm having trouble keeping track of what happened where anymore. 

The first is that if it matches, it will match the second user in the db every time and update their user_data and messages instead of properly pulling the correct user or identifying that they are not in the db. The database never grew past the first two people to chat.

Second, and I know this is more of me being a Js newbie, but I can't push the new message onto the messages result, instead, it sets the msgs var as the integer 2. *EDIT: well, this was a problem.. it seems to be working fine now.*

Finally, I'm unsure of how to use db.where properly. In the docs it shows a few different ways, but they all use statically typed examples. How do I properly do this with dynamic variables (such as username/channel). I have a feeling that is the root or at least part of the issue I'm having here, as in my console output I can see that it's matching when 'name' is not the same as user.username in the element it retrieves, it seems to be matching purely off the second condition of channel.

Apologies for my mistakes here, it's still very much all new territory for me and I hate clogging up the issues board with my own personal errors, but I really appreciate the help.

**Comment by celluj34** [comment]
I probably won't answer all of you questions, but here's a few things that I'm doing, maybe it can help:

I tried to use the db library from @Schmoopiie, but I couldn't deal with every function being a promise/callback, so I actually use lowdb (https://github.com/typicode/lowdb). It uses underscore/lowdash to interact with the database and is extremely straightforward in usage.

With that said, here's my highlight message function, which you could probably use to determine if a message contains a blacklisted keyword:


    var _s = require("underscore.string");
    ...
    server.locals.database = path.join(__dirname, "assets", "database.json");
    ...
    var database = lowdb(server.locals.database);
    var keywords = database("keywords");
    ...
    function highlightMessage(comment) {
        var casedComment = comment.toLowerCase();

        var highlight = keywords.find(function(keyword) {
            return _s.contains(casedComment, keyword.value.toLowerCase());
        });

        return !_.isUndefined(highlight);
    }

One more note, the `where` function is usually only used with a key/value pair, like `db.where({id: 3})`. Not sure about that lib, but in lowdb you would use the `filter` command.

Let me know if some of that helps, or you have more questions!

**Comment by casefc3s** [comment]
Thanks for the feedback/alternative suggestion, however I would prefer to utilize as many modules as possible from one author as it's more likely that they'll already be integrated and work well together. It's one of those things where I know my inexperience in matters would likely make adapting to a new library much harder than it needs to be. If it winds up that I simply cannot do what I want to do, then I will definitely consider expanding my horizons!

I noticed that by removing the conditional {name: user.username} && {chan: channel} in the db.where call to just {name: user.username} that all is well and fine, and it correctly pulls the appropriate table/element (or whatever it's called in proper database-speak). The only issue there is that I'd like to keep separate records for users if they're around on more than one channel (like myself).

I tried parsing it out with the alternative syntax to see if I could "trick" it into taking dynamic variables, i.e. db.where('userdata', ('@name == ' + n + ' && @chan == ' + channel))[...] but that didn't work either.

I suppose the other way to do it is to store an array of channel data in each name's entry, which I will attempt now!
